src/Moq/AmbientObserver.cs

```diff
@@ -0,0 +1,208 @@
+	/// </summary>
+	/// <remarks>
+	///   <para>
+	///     This component requires the active cooperation of the respective subsystems.
+	///     That is, invoked matchers and mocks call into <see cref="OnMatch(Match)"/>
+	///     or <see cref="OnInvocation(Mock, Invocation)"/> if an ambient observer is
+	///     active on the current thread.
+	///   </para>
+	///   <para>
+	///     This gets used in Moq's API to work around certain limitations of what kind
+	///     of constructs the Roslyn compilers allow in in-source LINQ expression trees
+	///     (e.g., assignment and event (un-)subscription are forbidden). Instead of
+	///     letting user code provide a LINQ expression tree, Moq accepts a normal lambda.
+	///     While a lambda cannot be directly inspected like a LINQ expression tree, we
+	///     can instantiate an <see cref="AmbientObserver"/>, execute the lambda, and then
+	///     check with the observer what invocations happened; and from there, we can
+	///     "reverse-engineer" a LINQ expression tree (with some loss of accuracy).
+	///   </para>
+	/// </remarks>
+	internal sealed class AmbientObserver : IDisposable
+	{
+		[ThreadStatic]
+		private static AmbientObserver current;
+
+		public static bool IsActive(out AmbientObserver observer)
+		{
+			var current = AmbientObserver.current;
+
+			observer = current;
+			return current != null;
+		}
+
+		private List<Observation> observations;
+
+		public AmbientObserver()
+		{
+			current = this;
+		}
+
+		public void Dispose()
+		{
+			if (this.observations != null)
+			{
+				for (var i = this.observations.Count - 1; i >= 0; --i)
+				{
+					this.observations[i].Dispose();
+				}
+			}
+
+			current = null;
+		}
+
+		/// <summary>
+		///   Adds the specified mock invocation as an observation.
+		/// </summary>
+		public void OnInvocation(Mock mock, Invocation invocation)
+		{
+			if (this.observations == null)
+			{
+				this.observations = new List<Observation>();
+			}
+
+			observations.Add(new InvocationObservation(mock, invocation));
+		}
+
+		/// <summary>
+		///   Adds the specified <see cref="Match"/> as an observation.
+		/// </summary>
+		public void OnMatch(Match match)
+		{
+			if (this.observations == null)
+			{
+				this.observations = new List<Observation>();
+			}
+
+			this.observations.Add(new MatchObservation(match));
+		}
+
+		/// <summary>
+		///   Checks whether the last observed thing was a mock invocation.
+		///   If <see langword="true"/>, details about that mock invocation are provided via the <see langword="out"/> parameters.
+		/// </summary>
+		/// <param name="mock">The <see cref="Mock"/> on which an invocation was observed.</param>
+		/// <param name="invocation">The observed <see cref="Invocation"/>.</param>
+		/// <param name="matches">The <see cref="Match"/>es that were observed just before the invocation.</param>
+		public bool LastObservationWasMockInvocation(out Mock mock, out Invocation invocation, out Matches matches)
```

Comment:
Name mismatch, `Invocation` in `OnInvocation` vs `MockInvocation` in `LastObservationWasMockInvocation`. It would be good to decide on one naming scheme.